File: android/app/src/main/kotlin/com/example/meal_planner/NativeAdFactory.kt
package com.example.meal_planner

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import com.google.android.gms.ads.nativead.NativeAd
import com.google.android.gms.ads.nativead.NativeAdView
import io.flutter.plugins.googlemobileads.GoogleMobileAdsPlugin

class GlassCardNativeAdFactory(private val context: Context) :
    GoogleMobileAdsPlugin.NativeAdFactory {

    override fun createNativeAd(
        nativeAd: NativeAd,
        customOptions: MutableMap<String, Any>?
    ): NativeAdView {
        val adView = LayoutInflater.from(context)
            .inflate(R.layout.native_ad_glass_card, null) as NativeAdView

        val headlineView = adView.findViewById<TextView>(R.id.ad_headline)
        headlineView.text = nativeAd.headline
        adView.headlineView = headlineView

        val bodyView = adView.findViewById<TextView>(R.id.ad_body)
        bodyView.text = nativeAd.body ?: ""
        bodyView.visibility = if (nativeAd.body != null) View.VISIBLE else View.GONE
        adView.bodyView = bodyView

        val iconView = adView.findViewById<ImageView>(R.id.ad_icon)
        if (nativeAd.icon != null) {
            iconView.setImageDrawable(nativeAd.icon!!.drawable)
            iconView.visibility = View.VISIBLE
        } else {
            iconView.visibility = View.GONE
        }
        adView.iconView = iconView

        adView.setNativeAd(nativeAd)
        return adView
    }
}

File: ios/Runner/NativeAdFactory.swift
import Foundation
import google_mobile_ads
import GoogleMobileAds

class GlassCardNativeAdFactory: FLTNativeAdFactory {
    func createNativeAd(
        _ nativeAd: GADNativeAd,
        customOptions: [String: Any]? = nil
    ) -> GADNativeAdView? {
        let nibObjects = Bundle.main.loadNibNamed("NativeAdView", owner: nil, options: nil)
        guard let adView = nibObjects?.first as? GADNativeAdView else {
            return nil
        }

        (adView.headlineView as? UILabel)?.text = nativeAd.headline
        (adView.bodyView as? UILabel)?.text = nativeAd.body
        adView.bodyView?.isHidden = nativeAd.body == nil

        (adView.iconView as? UIImageView)?.image = nativeAd.icon?.image
        adView.iconView?.isHidden = nativeAd.icon == nil

        adView.nativeAd = nativeAd
        return adView
    }
}
